src/main/java/jenkins/scm/api/trait/SCMSourceContext.java

```diff
@@ -0,0 +1,260 @@
+ * {@code return Collections.unmodifiableList(theList);} rather than the concurrency safe
+ * {@code return Collections.unmodifiableList(new ArrayList<>(theList));}
+ * </li>
+ * </ul>
+ *
+ * @param <C> the type of {@link SCMSourceContext}
+ * @param <R> the type of {@link SCMSourceRequest} produced by this context.
+ * @since 2.2.0
+ */
+public abstract class SCMSourceContext<C extends SCMSourceContext<C, R>, R extends SCMSourceRequest> {
+    /**
+     * The criteria.
+     */
+    @NonNull
+    private final List<SCMSourceCriteria> criteria = new ArrayList<SCMSourceCriteria>();
+    /**
+     * The pre-filters (i.e. filters that are independent of the {@link SCMSourceRequest}).
+     */
+    @NonNull
+    private final List<SCMHeadPrefilter> prefilters = new ArrayList<SCMHeadPrefilter>();
+    /**
+     * The {@link SCMSourceRequest} dependent filters.
+     */
+    @NonNull
+    private final List<SCMHeadFilter> filters = new ArrayList<SCMHeadFilter>();
+    /**
+     * The authorities.
+     */
+    @NonNull
+    private final List<SCMHeadAuthority<?, ?, ?>> authorities = new ArrayList<SCMHeadAuthority<?, ?, ?>>();
+    /**
+     * The observer.
+     */
+    @NonNull
+    private SCMHeadObserver observer;
+
+    /**
+     * Constructor.
+     *
+     * @param criteria (optional) criteria.
+     * @param observer the {@link SCMHeadObserver}.
+     */
+    public SCMSourceContext(@CheckForNull SCMSourceCriteria criteria, @NonNull SCMHeadObserver observer) {
+        withCriteria(criteria);
+        this.observer = observer;
+    }
+
+    /**
+     * Returns the (possibly empty) list of {@link SCMHeadAuthority} instances that can define trust.
+     *
+     * @return the (possibly empty) list of {@link SCMHeadAuthority} instances that can define trust.
+     */
+    @NonNull
+    public final List<SCMHeadAuthority<?, ?, ?>> authorities() {
+        return Collections.unmodifiableList(authorities);
+    }
+
+    /**
+     * Returns the (possibly empty) list of criteria that must be met by a candidate {@link SCMHead}.
+     *
+     * @return the (possibly empty) list of criteria that must be met by a candidate {@link SCMHead}.
+     */
+    @NonNull
+    public final List<SCMSourceCriteria> criteria() {
+        return Collections.unmodifiableList(criteria);
+    }
+
+    /**
+     * Returns the (possibly empty) list of {@link SCMSourceRequest} dependent filters.
+     *
+     * @return the (possibly empty) list of {@link SCMSourceRequest} dependent filters.
+     */
+    @NonNull
+    public final List<SCMHeadFilter> filters() {
+        return Collections.unmodifiableList(filters);
+    }
+
+    /**
+     * Returns the (possibly empty) list of {@link SCMSourceRequest} independent pre-filters.
+     *
+     * @return the (possibly empty) list of {@link SCMSourceRequest} independent pre-filters.
+     */
+    @NonNull
+    public final List<SCMHeadPrefilter> prefilters() {
```

Comment:
`prefilter` is a word, therefore camel-case would be `nonSense`

<img width="611" alt="screen shot 2017-06-14 at 09 36 42" src="https://user-images.githubusercontent.com/209336/27123214-164c9bba-50e5-11e7-9289-2f666acf4dd8.png">